docs/proposals/20210305-sidecarset-hotupgrade.md

```diff
@@ -0,0 +1,329 @@
+- hotUpgradeEmptyImage: when upgradeType=HotUpgrade, user needs to provide an empty container for hot upgrades
+
+**Note: hotUpgradeEmptyImage has the same configuration as the sidecar container, for example: command, lifecycle, probe, etc, but it doesn't do anything.**
+
+Hot upgrade consists of the following two processes:
+- inject hot upgrade sidecar containers
+- in-place hot upgrade sidecar container
+
+### Inject Containers
+
+When the sidecar container upgradeStrategy=HotUpgrade, the SidecarSet Webhook will inject two containers into the Pod:
+1. {sidecarContainer.name}-1: as shown in the figure below: envoy-1, the container run the actual working sidecar container, such as envoy:1.16.0
+2. {sidecarContainer.name}-2: as shown in the figure below: envoy-2, the container run the hot upgrade empty container, and it doesn't have to deal with any real logic, as long as it stays in place, such as empty:1.0
+
+![sidecarset hotupgrade_injection](../../docs/img/sidecarset_hotupgrade_injection.png)
+
+### Hot Upgrade
+
+The SidecarSet Controller breaks down the hot upgrade process of the sidecar container into three steps:
+1. Upgrade: upgrade the empty container to the new version of the sidecar container, such as envoy-2.Image = envoy:1.17.0
+2. Migration: the process completes the state migration of stateful container, which needs to be done by the business itself according to its own container characteristics.
+PostStartHook completes the migration of the above process.
+(**Note: PostStartHook must block during the migration, and exit when migration complete.**)
+3. Reset: the step resets the old version sidecar container into empty container, such as envoy-1.Image = empty:1.0
+
+The above is the complete hot upgrade process. If a Pod needs to be hot upgraded several times, the above three steps can be repeated.
+
+![sidecarset hotupgrade](../../docs/img/sidecarset_hotupgrade.png)
+
+#### Migration Mechanism
+
+SidecarSet Hot upgrade provides an upgrade mechanism for stateful container. The specific state transfer process still needs to be implemented by the business itself.
+State migration is a very important step in the hot upgrade process, so i will describe some of the key points below:
+
+**1. Coordination Mechanism**
+
+The state migration process starts with a coordination between the two sidecar containers to decide the migration direction (sidecar-1 -> sidecar-2 or sidecar-2 -> sidecar-1).
+Since state migration is always from old -> new version. For this case OpenKruise injects the environment variable **SIDECARSET_VERSION** into the sidecar container by default.
+The value of SIDECARSET_VERSION is the resourceVersion of the corresponding SidecarSet, so the larger one is the newer version compared with the two SIDECARSET_VERSION.
+As shown below, envoy-2 is the new version of sidecar:
+```
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: envoy-1
+    env:
+    - name: SIDECARSET_VERSION
```

Comment:
env in a pod is immutable once created, we should link the env to an annotations using downward api